include/topk_queue.hpp

```diff
@@ -1,59 +1,49 @@
 #pragma once
+
 namespace ds2i {
 
 struct topk_queue {
-    explicit topk_queue(uint64_t k) : threshold(0), m_k(k) {}
+    using entry_type = std::pair<float, uint64_t>;
+
+    explicit topk_queue(uint64_t k) : m_threshold(0), m_k(k) { m_q.reserve(k + 1); }
 
-    topk_queue(const topk_queue &q) : threshold(q.threshold), m_k(q.m_k), m_q(q.m_q) {
+    topk_queue(const topk_queue &q) : m_threshold(q.m_threshold), m_k(q.m_k), m_q(q.m_q) {
         m_k = q.m_k;
-        threshold = q.threshold;
+        m_threshold = q.m_threshold;
     }
 
     topk_queue &operator=(const topk_queue &q) {
         m_k = q.m_k;
-        threshold = q.threshold;
+        m_threshold = q.m_threshold;
         return *this;
     }
 
+    [[nodiscard]] constexpr static auto min_heap_order(entry_type const &lhs,
+                                                       entry_type const &rhs) noexcept -> bool {
+        return lhs.first > rhs.first;
+    }
+
     bool insert(float score) { return insert(score, 0); }
 
     bool insert(float score, uint64_t docid) {
+        if (DS2I_UNLIKELY(score < m_threshold)) {
+            return false;
+        }
+        m_q.emplace_back(score, docid);
         if (m_q.size() < m_k) {
-            m_q.push_back(std::make_pair(score, docid));
-            std::push_heap(m_q.begin(),
-                           m_q.end(),
-                           [](std::pair<float, uint64_t> l, std::pair<float, uint64_t> r) {
-                               return l.first > r.first;
-                           });
-            threshold = m_q.front().first;
-            return true;
+            std::push_heap(m_q.begin(), m_q.end(), min_heap_order);
         } else {
-            if (score > threshold) {
-                std::pop_heap(m_q.begin(),
-                              m_q.end(),
-                              [](std::pair<float, uint64_t> l, std::pair<float, uint64_t> r) {
-                                  return l.first > r.first;
-                              });
-                m_q.back() = std::make_pair(score, docid);
-                std::push_heap(m_q.begin(),
-                               m_q.end(),
-                               [](std::pair<float, uint64_t> l, std::pair<float, uint64_t> r) {
-                                   return l.first > r.first;
-                               });
-                threshold = m_q.front().first;
-                return true;
-            }
+            std::pop_heap(m_q.begin(), m_q.end(), min_heap_order);
+            m_q.pop_back();
         }
-        return false;
+        m_threshold = m_q.size() == m_k ? m_q.front().first : 0.0;
```

Comment:
I feel like we can avoid this check, since we are inserting there is something for sure. 
Shall we change it to `m_threshold = m_q.front().first`?

Reply:
Yes, you are right! I have updated it. Please let me know if this is what you mean.